src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+        // various purposes. These include:
+        //
+        // 1. The `shutdown` future above.
+        // 2. An RPC indicating the server should shut down
+        // 3. A period of inactivity (no requests serviced)
+        //
+        // These are all encapsulated wih the future that we're creating below.
+        // The `ShutdownOrInactive` indicates the RPC or the period of
+        // inactivity, and this is then select'd with the `shutdown` future
+        // passed to this function.
+        let handle = core.handle();
+        let shutdown_idle = ShutdownOrInactive {
+            rx: rx,
+            timeout: Timeout::new(timeout, &handle)?,
+            handle: handle.clone(),
+            timeout_dur: timeout,
+        };
+
+        let server = server.select(shutdown_idle);
+        let shutdown = shutdown.map_err(|()| {
+            io::Error::new(io::ErrorKind::Other, "shutdown signal failed")
+        });
+        let server = server.map(|_| ()).map_err(|e| e.0).select(shutdown);
+        core.run(server)
+            .map_err(|p| p.0)?;
+
+        // Once our server has shut down either due to inactivity or a manual
+        // request we still need to give a bit of time for all active
+        // connections to finish. This `wait` future will resolve once all
+        // instances of `SccacheService` have been dropped.
+        //
+        // Note that we cap the amount of time this can take, however, as we
+        // don't want to wait *too* long.
+        core.run(wait.select(Timeout::new(Duration::new(1, 0), &handle)?))
+            .map_err(|p| p.0)?;
+
+        Ok(())
+    }
+}
 
-    /// After this number of milliseconds with no client requests, shut down.
-    idle_timeout: u64,
+/// Service implementation for sccache
+#[derive(Clone)]
+struct SccacheService<C: CommandCreatorSync> {
+    /// Server statistics.
+    stats: Rc<RefCell<ServerStats>>,
 
-    /// A `Timeout` handle for the server idle shutdown timeout.
-    timeout: Option<Timeout>,
+    /// Cache storage.
+    storage: Arc<Storage>,
 
     /// A cache of known compiler info.
-    compilers: HashMap<String, Option<Compiler>>,
+    compilers: Rc<RefCell<HashMap<String, Option<Compiler>>>>,
 
     /// True if all compiles should be forced, ignoring existing cache entries.
     ///
     /// This can be controlled with the `SCCACHE_RECACHE` environment variable.
     force_recache: bool,
 
+    /// Thread pool to execute work in
+    pool: CpuPool,
+
+    /// Handle to the event loop that we're running on.
+    handle: Handle,
+
     /// An object for creating commands.
     ///
     /// This is mostly useful for unit testing, where we
     /// can mock this out.
     creator: C,
+
+    /// Message channel used to learn about requests received by this server.
+    tx: mpsc::Sender<ServerMessage>,
+
+    /// Information tracking how many services are active.
+    info: ActiveInfo,
+}
+
+type SccacheRequest = Message<ClientRequest, Body<(), io::Error>>;
```

Comment:
I don't love the `streaming::Message` type, honestly. It feels very HTTP-centric. It would be nicer to be able to just say "a client request is a single `ClientRequest`" here, and then "a server response is one or more `ServerResponse`s" below without the arbitrary head/body distinction.

Maybe that's something we can make easier in `tokio-proto`.